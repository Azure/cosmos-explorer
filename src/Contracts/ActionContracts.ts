/**
 * Data Explorer tab kinds
 */
export enum TabKind {
  SQLDocuments,
  MongoDocuments,
  MongoSchema,
  TableEntities,
  Graph,
  SQLQuery,
  ScaleSettings,
}

/**
 * Data Explorer pane kinds
 */
export enum PaneKind {
  AddCollection,
  CassandraAddCollection,
  DeleteCollection,
  DeleteDatabase,
  GlobalSettings,
  AdHocAccess,
  SwitchDirectory,
}

/**
 * Parameters to pass to DataExplorer in order to have it perform a given action
 */
export interface DataExplorerAction {
  actionType: ActionType | string;
}

/**
 * Open tab action
 */
export interface OpenTab extends DataExplorerAction {
  tabKind: TabKind | string;
}

/**
 * Open collection tab action
 */
export interface OpenCollectionTab extends OpenTab {
  databaseResourceId: string;
  collectionResourceId: string;
}

/**
 * Open query tab action
 */
export interface OpenQueryTab extends OpenCollectionTab {
  query: QueryInfo;
}

/**
 * Query info will be looked at in the order given in this interface (i.e. if text is provided, everything else is ignored)
 */
export interface QueryInfo {
  text?: string;
  partitionKeys?: string[];
}

/**
 * Open pane action
 */
export interface OpenPane extends DataExplorerAction {
  paneKind: PaneKind | string;
}

export interface OpenSampleNotebook extends DataExplorerAction {
  path: string;
}

/**
 * The types of actions that the DataExplorer supports performing upon opening.
 */
export enum ActionType {
  OpenTab,
  OpenCollectionTab,
  OpenPane,
  TransmitCachedData,
  OpenSampleNotebook,
}
