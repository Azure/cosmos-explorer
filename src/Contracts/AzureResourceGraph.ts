export interface QueryRequestOptions {
  $skipToken?: string;
  $top?: number;
  allowPartialScopes: boolean;
  subscriptions?: string[];
}

export interface QueryResponse {
  $skipToken: string;
  count: number;
  data: any;
  resultTruncated: boolean;
  totalRecords: number;
}
