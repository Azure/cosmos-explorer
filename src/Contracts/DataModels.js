"use strict";
exports.__esModule = true;
exports.SparkClusterEndpointKind = exports.ConflictResolutionMode = exports.ApiKind = void 0;
var ApiKind;
(function (ApiKind) {
    ApiKind[ApiKind["SQL"] = 0] = "SQL";
    ApiKind[ApiKind["MongoDB"] = 1] = "MongoDB";
    ApiKind[ApiKind["Table"] = 2] = "Table";
    ApiKind[ApiKind["Cassandra"] = 3] = "Cassandra";
    ApiKind[ApiKind["Graph"] = 4] = "Graph";
    ApiKind[ApiKind["MongoDBCompute"] = 5] = "MongoDBCompute";
})(ApiKind = exports.ApiKind || (exports.ApiKind = {}));
var ConflictResolutionMode;
(function (ConflictResolutionMode) {
    ConflictResolutionMode["Custom"] = "Custom";
    ConflictResolutionMode["LastWriterWins"] = "LastWriterWins";
})(ConflictResolutionMode = exports.ConflictResolutionMode || (exports.ConflictResolutionMode = {}));
var SparkClusterEndpointKind;
(function (SparkClusterEndpointKind) {
    SparkClusterEndpointKind["SparkUI"] = "SparkUI";
    SparkClusterEndpointKind["HistoryServerUI"] = "HistoryServerUI";
    SparkClusterEndpointKind["Livy"] = "Livy";
})(SparkClusterEndpointKind = exports.SparkClusterEndpointKind || (exports.SparkClusterEndpointKind = {}));
