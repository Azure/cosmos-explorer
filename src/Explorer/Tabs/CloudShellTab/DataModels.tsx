/**
 * Copyright (c) Microsoft Corporation.  All rights reserved.
 */

export const enum OsType {
    Linux = "linux",
    Windows = "windows"
}

export const enum ShellType {
    Bash = "bash",
    PowerShellCore = "pwsh"
}
  
export const enum NetworkType {
    Default = "Default",
    Isolated = "Isolated"
}

export const enum SessionType {
    Mounted = "Mounted",
    Ephemeral = "Ephemeral"
}

export type Settings = {
    location: string;
    sessionType: SessionType;
    osType: OsType;
};

export type ProvisionConsoleResponse = {
    properties: {
        osType: OsType;
        provisioningState: string;
        uri: string;
    };
};

export type Authorization = {
    token: string;
};

export type ConnectTerminalResponse = {
    id: string;
    idleTimeout: string;
    rootDirectory: string;
    socketUri: string;
    tokenUpdated: boolean;
};


