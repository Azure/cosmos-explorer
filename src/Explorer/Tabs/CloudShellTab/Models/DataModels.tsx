export const enum OsType {
  Linux = "linux",
  Windows = "windows",
}

export const enum ShellType {
  Bash = "bash",
  PowerShellCore = "pwsh",
}

export const enum NetworkType {
  Default = "Default",
  Isolated = "Isolated",
}

/**
 * Azure CloudShell session types:
 * - Mounted: Sessions with persistent storage via an Azure File Share mount.
 *   Files and configurations are preserved between sessions, allowing for
 *   continuity of work across multiple CloudShell sessions.
 *
 * - Ephemeral: Temporary sessions without persistent storage.
 *   All files and changes are discarded when the session ends.
 *   These sessions start faster but don't retain user data.
 *
 * The session type affects resource allocation, startup time,
 * and whether user files/configurations persist between sessions.
 */
export const enum SessionType {
  Mounted = "Mounted",
  Ephemeral = "Ephemeral",
}

export type CloudShellSettings = {
  properties: UserSettingProperties;
};

export type UserSettingProperties = {
  networkType: string;
  preferredLocation: string;
  preferredOsType: OsType;
  preferredShellType: ShellType;
  userSubscription: string;
  sessionType: SessionType;
  vnetSettings: object;
};

export type ProvisionConsoleResponse = {
  properties: {
    osType: OsType;
    provisioningState: string;
    uri: string;
  };
};

export type Authorization = {
  token: string;
};

export type ConnectTerminalResponse = {
  id: string;
  idleTimeout: string;
  rootDirectory: string;
  socketUri: string;
  tokenUpdated: boolean;
};

export type ProviderAuthorization = {
  applicationId: string;
  roleDefinitionId: string;
};

export type ProviderResourceType = {
  resourceType: string;
  locations: string[];
  apiVersions: string[];
  defaultApiVersion?: string;
  capabilities?: string;
};

export type RegistrationState = "Registered" | "NotRegistered" | "Registering" | "Unregistering";
export type RegistrationPolicy = "RegistrationRequired" | "RegistrationOptional";

export type CloudShellProviderInfo = {
  id: string;
  namespace: string;
  authorizations?: ProviderAuthorization[];
  resourceTypes: ProviderResourceType[];
  registrationState: RegistrationState;
  registrationPolicy: RegistrationPolicy;
};
