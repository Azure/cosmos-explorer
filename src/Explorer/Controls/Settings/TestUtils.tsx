import ko from "knockout";
import * as DataModels from "../../../Contracts/DataModels";
import * as ViewModels from "../../../Contracts/ViewModels";
import Explorer from "../../Explorer";

export const container = new Explorer();

export const collection = {
  container: container,
  databaseId: "test",
  id: ko.observable<string>("test"),
  defaultTtl: ko.observable<number>(5),
  analyticalStorageTtl: ko.observable<number>(undefined),
  indexingPolicy: ko.observable<DataModels.IndexingPolicy>({
    automatic: true,
    indexingMode: "consistent",
    includedPaths: [],
    excludedPaths: [],
  }),
  rawDataModel: {
    uniqueKeyPolicy: {
      uniqueKeys: [
        {
          paths: ["/id"],
        },
      ],
    },
  },
  usageSizeInKB: ko.observable(100),
  offer: ko.observable<DataModels.Offer>({
    autoscaleMaxThroughput: undefined,
    manualThroughput: 10000,
    minimumThroughput: 6000,
    id: "offer",
    offerReplacePending: false,
  }),
  conflictResolutionPolicy: ko.observable<DataModels.ConflictResolutionPolicy>(
    {} as DataModels.ConflictResolutionPolicy,
  ),
  changeFeedPolicy: ko.observable<DataModels.ChangeFeedPolicy>({} as DataModels.ChangeFeedPolicy),
  geospatialConfig: ko.observable<DataModels.GeospatialConfig>({} as DataModels.GeospatialConfig),
  getDatabase: () => {
    return;
  },
  partitionKey: {
    paths: [],
    kind: "hash",
    version: 2,
  },
  partitionKeyProperties: ["partitionKey"],
  computedProperties: ko.observable<DataModels.ComputedProperties>([
    {
      name: "queryName",
      query: "query",
    },
  ]),
  vectorEmbeddingPolicy: ko.observable<DataModels.VectorEmbeddingPolicy>({} as DataModels.VectorEmbeddingPolicy),
  fullTextPolicy: ko.observable<DataModels.FullTextPolicy>({} as DataModels.FullTextPolicy),
  materializedViews: ko.observable<DataModels.MaterializedView[]>([
    { id: "view1", _rid: "rid1" },
    { id: "view2", _rid: "rid2" },
  ]),
  materializedViewDefinition: ko.observable<DataModels.MaterializedViewDefinition>({
    definition: "SELECT * FROM c WHERE c.id = 1",
    sourceCollectionId: "source1",
    sourceCollectionRid: "rid123",
  }),
  readSettings: () => {
    return;
  },
} as unknown as ViewModels.Collection;
