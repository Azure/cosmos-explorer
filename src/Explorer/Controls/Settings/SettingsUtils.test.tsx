import ko from "knockout";
import * as DataModels from "../../../Contracts/DataModels";
import * as ViewModels from "../../../Contracts/ViewModels";
import {
  getMongoIndexType,
  getMongoIndexTypeText,
  getMongoNotification,
  getPartitionKeyName,
  getPartitionKeyPlaceHolder,
  getPartitionKeySubtext,
  getPartitionKeyTooltipText,
  getSanitizedInputValue,
  getTabTitle,
  hasDatabaseSharedThroughput,
  isDirty,
  isIndexTransforming,
  MongoIndexTypes,
  MongoNotificationType,
  MongoWildcardPlaceHolder,
  parseConflictResolutionMode,
  parseConflictResolutionProcedure,
  SettingsV2TabTypes,
  SingleFieldText,
  WildcardText,
} from "./SettingsUtils";
import { collection } from "./TestUtils";

describe("SettingsUtils", () => {
  it("hasDatabaseSharedThroughput", () => {
    expect(hasDatabaseSharedThroughput(collection)).toEqual(undefined);

    const newCollection = { ...collection };
    newCollection.getDatabase = () => {
      return {
        nodeKind: undefined,
        rid: undefined,
        container: undefined,
        self: "",
        id: ko.observable(""),
        collections: ko.observableArray(undefined),
        offer: ko.observable(undefined),
        isDatabaseExpanded: ko.observable(false),
        isDatabaseShared: ko.computed(() => true),
        selectedSubnodeKind: ko.observable(undefined),
        expandDatabase: undefined,
        collapseDatabase: undefined,
        loadCollections: undefined,
        findCollectionWithId: undefined,
        openAddCollection: undefined,
        readSettings: undefined,
        onSettingsClick: undefined,
        loadOffer: undefined,
      } as ViewModels.Database;
    };
    newCollection.offer(undefined);
    expect(hasDatabaseSharedThroughput(newCollection)).toEqual(true);
  });

  describe("parseConflictResolutionMode", () => {
    it("parses valid modes correctly", () => {
      expect(parseConflictResolutionMode("custom")).toEqual(DataModels.ConflictResolutionMode.Custom);
      expect(parseConflictResolutionMode("lastwriterwins")).toEqual(DataModels.ConflictResolutionMode.LastWriterWins);
      expect(parseConflictResolutionMode("Custom")).toEqual(DataModels.ConflictResolutionMode.Custom);
      expect(parseConflictResolutionMode("CUSTOM")).toEqual(DataModels.ConflictResolutionMode.Custom);
      expect(parseConflictResolutionMode("LastWriterWins")).toEqual(DataModels.ConflictResolutionMode.LastWriterWins);
    });

    it("handles empty/undefined input", () => {
      expect(parseConflictResolutionMode(undefined)).toBeUndefined();
      expect(parseConflictResolutionMode(null)).toBeUndefined();
      expect(parseConflictResolutionMode("")).toBeUndefined();
    });

    it("defaults to LastWriterWins for invalid inputs", () => {
      expect(parseConflictResolutionMode("invalid")).toEqual(DataModels.ConflictResolutionMode.LastWriterWins);
      expect(parseConflictResolutionMode("123")).toEqual(DataModels.ConflictResolutionMode.LastWriterWins);
    });
  });

  describe("parseConflictResolutionProcedure", () => {
    it("extracts procedure name from valid paths", () => {
      expect(parseConflictResolutionProcedure("/dbs/db/colls/coll/sprocs/conflictResSproc")).toEqual(
        "conflictResSproc",
      );
      expect(parseConflictResolutionProcedure("conflictResSproc")).toEqual("conflictResSproc");
      expect(parseConflictResolutionProcedure("/dbs/mydb/colls/mycoll/sprocs/myProc")).toEqual("myProc");
    });

    it("handles empty/undefined input", () => {
      expect(parseConflictResolutionProcedure(undefined)).toBeUndefined();
      expect(parseConflictResolutionProcedure(null)).toBeUndefined();
      expect(parseConflictResolutionProcedure("")).toBeUndefined();
    });

    it("handles invalid path formats", () => {
      expect(parseConflictResolutionProcedure("/invalid/path")).toBeUndefined();
      expect(parseConflictResolutionProcedure("/dbs/db/colls/coll/wrongtype/name")).toBeUndefined();
    });
  });

  describe("isDirty", () => {
    const indexingPolicy = {
      automatic: true,
      indexingMode: "consistent",
      includedPaths: [],
      excludedPaths: [],
    } as DataModels.IndexingPolicy;

    describe("primitive types", () => {
      it("handles strings", () => {
        expect(isDirty("baseline", "baseline")).toBeFalsy();
        expect(isDirty("baseline", "current")).toBeTruthy();
        expect(isDirty("", "")).toBeFalsy();
        expect(isDirty("test", "")).toBeTruthy();
      });

      it("handles numbers", () => {
        expect(isDirty(0, 0)).toBeFalsy();
        expect(isDirty(1, 1)).toBeFalsy();
        expect(isDirty(0, 1)).toBeTruthy();
        expect(isDirty(-1, 1)).toBeTruthy();
      });

      it("handles booleans", () => {
        expect(isDirty(true, true)).toBeFalsy();
        expect(isDirty(false, false)).toBeFalsy();
        expect(isDirty(true, false)).toBeTruthy();
      });

      it("handles undefined and null", () => {
        expect(isDirty(undefined, undefined)).toBeFalsy();
        expect(isDirty(null, null)).toBeFalsy();
        expect(isDirty(undefined, null)).toBeTruthy();
        expect(isDirty(undefined, "value")).toBeTruthy();
      });
    });

    describe("complex types", () => {
      it("handles indexing policy", () => {
        expect(isDirty(indexingPolicy, indexingPolicy)).toBeFalsy();
        expect(isDirty(indexingPolicy, { ...indexingPolicy, automatic: false })).toBeTruthy();
        expect(isDirty(indexingPolicy, { ...indexingPolicy, includedPaths: ["/path"] })).toBeTruthy();
      });

      it("handles array type policies", () => {
        const computedProperties: DataModels.ComputedProperties = [
          { name: "prop1", query: "SELECT * FROM c" },
          { name: "prop2", query: "SELECT * FROM c" }
        ];
        const otherProperties: DataModels.ComputedProperties = [
          { name: "prop1", query: "SELECT * FROM c" }
        ];
        expect(isDirty(computedProperties, computedProperties)).toBeFalsy();
        expect(isDirty(computedProperties, otherProperties)).toBeTruthy();
      });
    });

    describe("type mismatch handling", () => {
      it("throws error for mismatched types", () => {
        expect(() => isDirty("string", 123)).toThrow("current and baseline values are not of the same type");
        expect(() => isDirty(true, "true")).toThrow("current and baseline values are not of the same type");
        expect(() => isDirty(0, false)).toThrow("current and baseline values are not of the same type");
      });
    });
  });

  describe("getSanitizedInputValue", () => {
    const max = 100;

    it("handles empty or invalid inputs", () => {
      expect(getSanitizedInputValue("", max)).toEqual(0);
      expect(getSanitizedInputValue("abc", max)).toEqual(0);
      expect(getSanitizedInputValue("!@#", max)).toEqual(0);
      expect(getSanitizedInputValue(null, max)).toEqual(0);
      expect(getSanitizedInputValue(undefined, max)).toEqual(0);
    });

    it("handles valid inputs within max", () => {
      expect(getSanitizedInputValue("10", max)).toEqual(10);
      expect(getSanitizedInputValue("50", max)).toEqual(50);
      expect(getSanitizedInputValue("100", max)).toEqual(100);
    });

    it("handles inputs exceeding max", () => {
      expect(getSanitizedInputValue("101", max)).toEqual(100);
      expect(getSanitizedInputValue("999", max)).toEqual(100);
      expect(getSanitizedInputValue("1000000", max)).toEqual(100);
    });

    it("handles inputs without max constraint", () => {
      expect(getSanitizedInputValue("10")).toEqual(10);
      expect(getSanitizedInputValue("1000")).toEqual(1000);
      expect(getSanitizedInputValue("999999")).toEqual(999999);
    });

    it("handles negative numbers", () => {
      expect(getSanitizedInputValue("-10", max)).toEqual(-10);
      expect(getSanitizedInputValue("-999", max)).toEqual(-999);
    });
  });

  describe("getMongoIndexType", () => {
    it("correctly identifies single field indexes", () => {
      expect(getMongoIndexType(["Single"])).toEqual(MongoIndexTypes.Single);
      expect(getMongoIndexType(["field1"])).toEqual(MongoIndexTypes.Single);
      expect(getMongoIndexType(["name"])).toEqual(MongoIndexTypes.Single);
    });

    it("correctly identifies wildcard indexes", () => {
      expect(getMongoIndexType(["Wildcard.$**"])).toEqual(MongoIndexTypes.Wildcard);
      expect(getMongoIndexType(["field.$**"])).toEqual(MongoIndexTypes.Wildcard);
      expect(getMongoIndexType(["nested.path.$**"])).toEqual(MongoIndexTypes.Wildcard);
    });

    it("returns undefined for invalid or compound indexes", () => {
      expect(getMongoIndexType(["Key1", "Key2"])).toBeUndefined();
      expect(getMongoIndexType([])).toBeUndefined();
      expect(getMongoIndexType(undefined)).toBeUndefined();
    });
  });

  describe("getMongoIndexTypeText", () => {
    it("returns correct text for single field indexes", () => {
      expect(getMongoIndexTypeText(MongoIndexTypes.Single)).toEqual(SingleFieldText);
    });

    it("returns correct text for wildcard indexes", () => {
      expect(getMongoIndexTypeText(MongoIndexTypes.Wildcard)).toEqual(WildcardText);
    });
  });

  describe("getMongoNotification", () => {
    const singleIndexDescription = "sampleKey";
    const wildcardIndexDescription = "sampleKey.$**";

    describe("type validation", () => {
      it("returns warning when type is missing", () => {
        const notification = getMongoNotification(singleIndexDescription, undefined);
        expect(notification.message).toEqual("Please select a type for each index.");
        expect(notification.type).toEqual(MongoNotificationType.Warning);
      });

      it("returns undefined for valid type and description combinations", () => {
        expect(getMongoNotification(singleIndexDescription, MongoIndexTypes.Single)).toBeUndefined();
        expect(getMongoNotification(wildcardIndexDescription, MongoIndexTypes.Wildcard)).toBeUndefined();
      });
    });

    describe("field name validation", () => {
      it("returns error when field name is empty", () => {
        const notification = getMongoNotification("", MongoIndexTypes.Single);
        expect(notification.message).toEqual("Please enter a field name.");
        expect(notification.type).toEqual(MongoNotificationType.Error);

        const whitespaceNotification = getMongoNotification("   ", MongoIndexTypes.Single);
        expect(whitespaceNotification.message).toEqual("Please enter a field name.");
        expect(whitespaceNotification.type).toEqual(MongoNotificationType.Error);
      });

      it("returns error when wildcard index is missing $** pattern", () => {
        const notification = getMongoNotification(singleIndexDescription, MongoIndexTypes.Wildcard);
        expect(notification.message).toEqual(
          "Wildcard path is not present in the field name. Use a pattern like " + MongoWildcardPlaceHolder,
        );
        expect(notification.type).toEqual(MongoNotificationType.Error);
      });
    });

    it("handles undefined field name", () => {
      const notification = getMongoNotification(undefined, MongoIndexTypes.Single);
      expect(notification.message).toEqual("Please enter a field name.");
      expect(notification.type).toEqual(MongoNotificationType.Error);
    });
  });
  it("isIndexingTransforming", () => {
    expect(isIndexTransforming(undefined)).toBeFalsy();
    expect(isIndexTransforming(0)).toBeTruthy();
    expect(isIndexTransforming(90)).toBeTruthy();
    expect(isIndexTransforming(100)).toBeFalsy();
  });

  describe("getTabTitle", () => {
    it("returns correct titles for each tab type", () => {
      expect(getTabTitle(SettingsV2TabTypes.ScaleTab)).toBe("Scale");
      expect(getTabTitle(SettingsV2TabTypes.ConflictResolutionTab)).toBe("Conflict Resolution");
      expect(getTabTitle(SettingsV2TabTypes.SubSettingsTab)).toBe("Settings");
      expect(getTabTitle(SettingsV2TabTypes.IndexingPolicyTab)).toBe("Indexing Policy");
      expect(getTabTitle(SettingsV2TabTypes.ComputedPropertiesTab)).toBe("Computed Properties");
      expect(getTabTitle(SettingsV2TabTypes.ContainerVectorPolicyTab)).toBe("Container Policies");
      expect(getTabTitle(SettingsV2TabTypes.ThroughputBucketsTab)).toBe("Throughput Buckets");
      expect(getTabTitle(SettingsV2TabTypes.DataMaskingTab)).toBe("Data Masking (Preview)");
    });

    it("handles partition key tab title based on fabric native", () => {
      // Assuming initially not fabric native
      expect(getTabTitle(SettingsV2TabTypes.PartitionKeyTab)).toBe("Partition Keys (preview)");
    });

    it("throws error for unknown tab type", () => {
      expect(() => getTabTitle(999 as SettingsV2TabTypes)).toThrow("Unknown tab 999");
    });
  });

  describe("partition key utils", () => {
    it("getPartitionKeyName returns correct name based on API type", () => {
      expect(getPartitionKeyName("Mongo")).toBe("Shard key");
      expect(getPartitionKeyName("SQL")).toBe("Partition key");
      expect(getPartitionKeyName("Mongo", true)).toBe("shard key");
      expect(getPartitionKeyName("SQL", true)).toBe("partition key");
    });

    it("getPartitionKeyTooltipText returns correct tooltip based on API type", () => {
      const mongoTooltip = getPartitionKeyTooltipText("Mongo");
      expect(mongoTooltip).toContain("shard key");
      expect(mongoTooltip).toContain("replica sets");

      const sqlTooltip = getPartitionKeyTooltipText("SQL");
      expect(sqlTooltip).toContain("partition key");
      expect(sqlTooltip).toContain("id is often a good choice");
    });

    it("getPartitionKeySubtext returns correct subtext", () => {
      expect(getPartitionKeySubtext(true, "SQL")).toBe(
        "For small workloads, the item ID is a suitable choice for the partition key.",
      );
      expect(getPartitionKeySubtext(true, "Mongo")).toBe(
        "For small workloads, the item ID is a suitable choice for the partition key.",
      );
      expect(getPartitionKeySubtext(false, "SQL")).toBe("");
      expect(getPartitionKeySubtext(true, "Other")).toBe("");
    });

    it("getPartitionKeyPlaceHolder returns correct placeholder based on API type", () => {
      expect(getPartitionKeyPlaceHolder("Mongo")).toBe("e.g., categoryId");
      expect(getPartitionKeyPlaceHolder("Gremlin")).toBe("e.g., /address");
      expect(getPartitionKeyPlaceHolder("SQL")).toBe("Required - first partition key e.g., /TenantId");
      expect(getPartitionKeyPlaceHolder("SQL", 0)).toBe("second partition key e.g., /UserId");
      expect(getPartitionKeyPlaceHolder("SQL", 1)).toBe("third partition key e.g., /SessionId");
      expect(getPartitionKeyPlaceHolder("Other")).toBe("e.g., /address/zipCode");
    });
  });
});
