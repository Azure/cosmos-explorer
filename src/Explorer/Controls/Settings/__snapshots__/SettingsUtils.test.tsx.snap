// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsUtils functions render 1`] = `
<Fragment>
  <span>
    Your 
    container
     throughput will automatically scale from
     
    <b>
      100
       RU/s (10% of max RU/s) -
       
      1000
       RU/s
    </b>
     
    based on usage. 
    <br />
    <br />
    After the first 
    10
     GB of data stored, the max RU/s will be automatically upgraded based on the new storage value.
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
       
      Learn more
    </a>
    .
  </span>
  <span>
    Your 
    database
     throughput will automatically scale from
     
    <b>
      100
       RU/s (10% of max RU/s) -
       
      1000
       RU/s
    </b>
     
    based on usage. 
    <br />
    <br />
    After the first 
    10
     GB of data stored, the max RU/s will be automatically upgraded based on the new storage value.
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
       
      Learn more
    </a>
    .
  </span>
  <span>
    Your 
    container
     throughput will automatically scale between 400 RU/s and 4,000 RU/s based on the workload needs, as long as your storage does not exceed 50GB. If your storage exceeds 50GB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Your 
    container
     throughput will automatically scale between 2,000 RU/s and 20,000 RU/s based on the workload needs, as long as your storage does not exceed 200GB. If your storage exceeds 200GB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Your 
    container
     throughput will automatically scale between 10,000 RU/s and 100,000 RU/s based on the workload needs, as long as your storage does not exceed 1TB. If your storage exceeds 1TB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Your 
    container
     throughput will automatically scale between 50,000 RU/s and 500,000 RU/s based on the workload needs, as long as your storage does not exceed 5TB. If your storage exceeds 5TB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Your 
    database
     throughput will automatically scale between 400 RU/s and 4,000 RU/s based on the workload needs, as long as your storage does not exceed 50GB. If your storage exceeds 50GB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Your 
    database
     throughput will automatically scale between 2,000 RU/s and 20,000 RU/s based on the workload needs, as long as your storage does not exceed 200GB. If your storage exceeds 200GB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Your 
    database
     throughput will automatically scale between 10,000 RU/s and 100,000 RU/s based on the workload needs, as long as your storage does not exceed 1TB. If your storage exceeds 1TB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Your 
    database
     throughput will automatically scale between 50,000 RU/s and 500,000 RU/s based on the workload needs, as long as your storage does not exceed 5TB. If your storage exceeds 5TB, we will upgrade the maximum (and minimum) throughput thresholds to the next available value. For more details, see
     
    <a
      href="https://aka.ms/cosmos-autoscale-info"
      rel="noreferrer"
      target="_blank"
    >
      documentation
    </a>
    .
  </span>
  <span>
    Estimated cost (
    RMB
    ):
     
    <b>
      ¥
      1.29
       hourly 
       / 
      ¥
      31.06
       daily 
       / 
      ¥
      944.60
       monthly
       
    </b>
    (
    2
     
    regions
    , 
    1000
    RU/s, 
    ¥
    0.00051
    /RU)
  </span>
  <span>
    Estimated monthly cost (
    RMB
    ):
     
    <b>
      ¥
      111.69
       - 
      ¥
      1116.90
       
    </b>
    (
    2
     
    regions
    , 
    100
     - 
    1000
     RU/s, 
    ¥
    0.000765
    /RU)
  </span>
  <span>
    The starting autoscale max RU/s will be determined by the system, based on the current manual throughput settings and storage of your resource. After autoscale has been enabled, you can change the max RU/s.
     
    <a
      href="https://aka.ms/cosmos-autoscale-migration"
    >
      Learn more
    </a>
  </span>
  <span>
    The system will automatically delete items based on the TTL value (in seconds) you provide, without needing a delete operation explicitly issued by a client application. For more information see,
     
    <a
      href="https://aka.ms/cosmos-db-ttl"
      rel="noreferrer"
      target="_blank"
    >
      Time to Live (TTL) in Azure Cosmos DB
    </a>
    .
  </span>
  <span>
    Changing the Indexing Policy impacts query results while the index transformation occurs. When a change is made and the indexing mode is set to consistent or lazy, queries return eventual results until the operation completes. For more information see,
     
    <a
      href="https://aka.ms/cosmosdb/modify-index-policy"
      rel="noreferrer"
      target="_blank"
    >
      Modifying Indexing Policies
    </a>
    .
  </span>
  <span>
    You are about to request an increase in throughput beyond the pre-allocated capacity. The service will scale out and increase throughput for the selected container. This operation will take 1-3 business days to complete. You can track the status of this request in Notifications.
  </span>
  <span>
    You are about to request an increase in throughput beyond the pre-allocated capacity. This operation will take some time to complete.
  </span>
  <span>
    The request to increase the throughput has successfully been submitted. This operation will take 1-3 business days to complete. View the latest status in Notifications.
    <br />
    Database: 
    sampleDb
    , Container: 
    sampleCollection
     
    , Current manual throughput: 1000 RU/s, Target manual throughput: 2000
  </span>
  <span>
    A request to increase the throughput is currently in progress. This operation will take some time to complete.
    <br />
    Database: 
    sampleDb
    , Container: 
    sampleCollection
     
    , Current manual throughput: 1000 RU/s, Target manual throughput: 2000
  </span>
  <span>
    A request to increase the throughput is currently in progress. This operation will take 1-3 business days to complete. View the latest status in Notifications.
    <br />
    Database: 
    sampleDb
    , Container: 
    sampleCollection
     
    , Current manual throughput: 1000 RU/s, Target manual throughput: 2000
  </span>
</Fragment>
`;
